contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,315 @@
+  // InodesUsed represents the inodes consumed on the device  
+  // +optional  
+  InodesUsed *uint64 `json:"inodesUsed,omitempty"`  
+}  
+```
+
+### Core Machine Info:
+
+In addition to providing metrics, cAdvisor also provides machine info.  While it is not neccessary to use this structure for reporting Machine Info, the following contains all of the information provided to the kubelet by cAdvisor, generally used at startup.
+
+The code that provides this data currently resides in cAdvisor.  I propose moving this to the kubelet as well.
+
+```go
+type CoreInfo struct {  
+  // MachineID reported by the node. For unique machine identification  
+  // in the cluster this field is prefered. Learn more from man(5)  
+  // machine-id: http://man7.org/linux/man-pages/man5/machine-id.5.html  
+  MachineID string `json:"machineID" protobuf:"bytes,1,opt,name=machineID"`  
+  // SystemUUID reported by the node. For unique machine identification  
+  // MachineID is prefered. This field is specific to Red Hat hosts  
+  // https://access.redhat.com/documentation/en-US/Red_Hat_Subscription_Management/1/html/RHSM/getting-system-uuid.html  
+  SystemUUID string `json:"systemUUID" protobuf:"bytes,2,opt,name=systemUUID"`  
+  // Boot ID reported by the node.  
+  BootID string `json:"bootID" protobuf:"bytes,3,opt,name=bootID"`  
+  // Kernel Version reported by the node from 'uname -r' (e.g. 3.16.0-0.bpo.4-amd64).  
+  KernelVersion string `json:"kernelVersion" protobuf:"bytes,4,opt,name=kernelVersion"`  
+  // OS Image reported by the node from /etc/os-release (e.g. Debian GNU/Linux 7 (wheezy)).  
+  OSImage string `json:"osImage" protobuf:"bytes,5,opt,name=osImage"`  
+  // Capacity represents the total resources of a node.  
+  // More info: http://kubernetes.io/docs/user-guide/persistent-volumes#capacity for more details.  
+  // +optional  
+  // ContainerRuntime Version reported by the node through runtime remote API (e.g. docker://1.5.0).  
+  ContainerRuntimeVersion string `json:"containerRuntimeVersion" protobuf:"bytes,6,opt,name=containerRuntimeVersion"`  
+  // Cloud provider the machine belongs to.  
+  CloudProvider CloudProvider `json:"cloud_provider"`  
+  // ID of cloud instance (e.g. instance-1) given to it by the cloud provider.  
+  InstanceID InstanceID `json:"instance_id"`  
+}  
+
+type CloudProvider string  
+
+const (  
+  GCE             CloudProvider = "GCE"  
+  AWS                           = "AWS"  
+  Azure                         = "Azure"  
+  Baremetal                     = "Baremetal"  
+  UnknownProvider               = "Unknown"  
+)
+```
+
+## Implementation Plan
+
+I will move all code pertaining to collection and processing of core metrics from cAdvisor into kubernetes.  
+I will vendor the new core metrics code back into cAdvisor. 
+I will modify volume stats collection so that it relies on this code.  
+I will modify the structure of stats collection code to be "On-Demand"   
```

Comment:
We need a mechanism to rate limit or control how frequently on-demand stat collection can occur.  Did you have thoughts here?  Even at existing 15s intervals, file system usage stats are very expensive to compute.